Add Crane 2.0 enhancement proposals (#288)
konveyor/enhancements: enhancements/crane-2.0/multi-stage-kustomize-transforms/README.md

---
title: multi-stage-kustomize-transforms
authors:
  - "@aufi"
reviewers:
  - "@rromannissen"
  - "@dymurray"
  - "@jwmatthews"
  - "@shawn-hurley"
  - "@pranavgaikwad"
approvers:
  - "@istein1"
  - "@aufi"
creation-date: 2026-05-05
last-updated: 2026-05-05
status: implementable
see-also:
  - "https://github.com/migtools/crane/issues/207"
  - "https://github.com/migtools/crane/blob/main/docs/transform.md"
---

# Inspectable Multi-Stage Transformations with Standard Kustomize Tools

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [x] Design details are appropriately documented from clear requirements
- [x] Test plan is defined
- [ ] User-facing documentation is created

## Open Questions

> 1. **Migration descriptor file vs. CLI flags + conventions (dev preferred):** Should Crane require a declarative descriptor file that specifies which transformers run and in what order, or is the current approach sufficient — CLI commands with flags (`--stage`) combined with the convention-based directory structure (`10_PluginName/`, `20_PluginName/`, `50_CustomChange/`)? The current approach keeps configuration minimal and co-located with the output. A future migration pipeline abstraction may introduce a migration plan file (including these descriptors) naturally, but it is unclear whether the transform descriptor file is needed at the single-tool level.

Reviewer: Can you provide examples of the intended input command for a multi stage transformation? Also related to that, how does the user know about the available plugins? How does the user define the order in which they are executed for a certain transformation?

Regarding the question itself, I think a migration descriptor approach is the best choice, as it would provide a standard file that could be versioned and stored in a repository to handle the repeatable transformation configuration of multiple resources at scale. For example, think about a resource transformation pipeline that has a stage that retrieves the transformation configuration from a repository (our migration descriptor) to then execute the transform command and store the adapted results on a different repository.

Reviewer 2: Sure, multistage commands:

(having exported data from source cluster already)

```
crane transform 
```
^ generates default Kubernetes plugin stage into transform/10_KubernetesPlugin

```
crane transform list-plugins
```
^ outputs available plugins list (default, crane-embeded and also custom from local dirs if available)

```
crane transform --stage 20_OpenShiftPlugin
```
^ creates a new stage with OpenShiftPlugin run after 10_Kubernetes (ordering by lexical order of stage name), with resources from previous stage output

```
crane transform --stage 50_CustomLabelsOrSomething # (no Plugin suffix)
```
^ creates a new stage with data from previous stage and kustomize skeleton files that changes nothing by default, but user can do their custom changes

```
crane apply
```
^ creates output directory with transformed manifests, note: in multistage, each new stage needs applied/materialized output of previous stage, so internally transform runs apply on previous stage to get input resources for a new one

More background on multistage transform with examples is in one of our doc PRs https://github.com/aufi/crane/blob/b53e2565959c54f9f677f2b622c433fce4f0a336/docs/transform-scenarios/03-multistage.md

Reviewer 2: On migration descriptor file, thanks for the example, it makes a perfect sense.

My/engineering _soft_ push-back comes from the approach to build crane as a set of small tools doing just their job as good as they can (unix-like style) and under influence of crane-runner project (https://github.com/migtools/crane-runner for pipeline based in-cluster migration orchestration and later potentially backend for UI). From this perspective, `crane transform` and `crane apply` are low-level commands, that (IMO) just get args and do their part of the job.

There is no a technical problem with implementing the descriptor file for transformations (crane transform could get a file with stages/plugins instead of single stage and generate those). The "dirty" part from my point of view is about having this descriptor file and then add (or extend it) to a "bigger" migration descriptor/plan file.

In the end (in later releases), I believe, there should be a single migration descriptor/plan file, that will be consumed by a single endpoint in crane (or related tooling) and it should cover all options needed for the full migration. Having a migration descriptor/migration plan file (that can contain src/dest clusters, transformers/plugins list, etc.) would fit better to an orchestration layer "above" those (`transform` and `apply`) commands.

This file will not be directly visible to `transform` or `apply` subcommands, that are just "cogs" in the migration process.

Reviewer 2: For better understading each other & discussion, there is a draft of possible migration descriptor / instructions file https://github.com/aufi/crane/tree/transform-instructions-file/instructions_file_notes#readme

Author: After discussing it we agreed on updating the transform with multiple stages generated by a single command and add simple transform descriptor file. Tracked in https://github.com/migtools/crane/issues/370

## Proposal

### User Stories

#### Story 1: Transformation with the Default Plugin

**As a** platform engineer migrating an application, **I want to** run `crane transform` to execute the default `10_KubernetesPlugin`, **So that** server-managed metadata is removed, generating standard Kustomize patches that I can easily inspect and debug.

* **Patch Generation:** The plugin automatically generates JSONPatch operations that Crane transparently converts into Kustomize patches, saving them in `transform/10_KubernetesPlugin/patches/` using a predictable, Git-friendly naming convention (`<namespace>--<group>-<version>--<Kind>--<name>.patch.yaml`).
* **Inspection:** At any time, I can preview the final cleaned YAML before applying it to the cluster by simply running `kubectl kustomize transform/10_KubernetesPlugin/`.
* **Debugging:** If I need to see exactly what went into the transformation or what came out, I can check the intermediate raw snapshots inside the `transform/.work/10_KubernetesPlugin/input/` and `output/` directories.

#### Story 2: Multi-Stage Transformation with Multiple Existing Plugins

> **As a** platform engineer performing a cross-platform migration (such as moving from OpenShift to vanilla Kubernetes), **I want to** run a multi-stage pipeline where each plugin has its own isolated directory (e.g., `10_KubernetesPlugin` and `20_OpenShiftPlugin`), **So that** I can accurately trace patch generation across plugins, inspect intermediate states, and debug unexpected changes without dealing with a single massive patch file.

Reviewer: This feels like design decisions (directory layout) bleeding into use cases.

Author: Resolving as there is no action required here.

* **Debugging:** If my manual edits to the `kustomization.yaml` are incorrect or malformed, running `crane apply` will fail and provide a reasonable error message for me to troubleshoot. If my modifications drift or I need to overwrite my dirty custom stage directory, I can force recreation using the `crane transform --force` flag. It is expected that the last stage is modified manually to be sure that the following stage will not be broken by the previous one.

### Implementation Details/Notes/Constraints

#### On-Disk Layout: Single-Stage

```text
transform/
  10_KubernetesPlugin/
    kustomization.yaml
    resources/
    patches/
```

#### On-Disk Layout: Multiple-Stages

```text
transform/
  10_KubernetesPlugin/
    kustomization.yaml
    resources/
    patches/
  20_OpenShiftPlugin/
    kustomization.yaml
    resources/
    patches/
  50_CustomModifications/
    kustomization.yaml
    resources/
```

Each stage consumes the `resources/` directory from the previous stage as input. The first stage consumes the export directory.

> #### CLI Changes

Reviewer: Crane is a new project to the Konveyor ecosystem. It would be convenient to explain what the full list of expected commands are after this enhancement, not only the ones that are new or have been changed.

Reviewer 2: Yes, thanks for suggestion, will update.

Reviewer 2: Added on lines below with https://github.com/konveyor/enhancements/pull/288/commits/33ec87c51885a1006597ba4f391dead32bb56ee0

> **`crane transform` — new flags:**

Reviewer: Can you explain what is the expected command and input when the user intends to perform a transformation chaining multiple plugins? I get the impression that the directory structure is being mixed as both input and output.

Reviewer 2: Correct, each stage contains `resources/` which is its input (that came from either export/ directory for the first stage or from the output of previous stage) and the result of given stage goes to following one (or to output/ directory for final stage when call `apply`), so right, inputs and outputs exist in within the transform directory. 

Reason for that each stage need to work on something real (if there is not yet output of previous stage, it is hard to generate new transformations on unknown data). This bring flexibility and isolation between stages. A limitation it brings is not recommended modification of "previous" stages, that could potentialy break following, already existing ones.

E.g. existing stage 20_... wants patch a resource, but someone does a manual change in already existing previous stage like 10_... that whiteouts such resource, then final `apply` will fail and the broken stage needs to be regenerated again with `transform`, easy for plugin-based stages, or fixed manually), so it make sense suggest users to manually modify just last existing stage (or pay attention to following stage update too).

More on multistage concepts and examples on https://github.com/migtools/crane/pull/307/changes#diff-5e44e00052da9942963c922d4e9f19a89db3d9da4e07d3fca9abec71d76266ce

Author: After discussing it we agreed on updating the transform with multiple stages generated by a single command and add simple transform descriptor file. Tracked in https://github.com/migtools/crane/issues/370

| Flag | Purpose |
|------|---------|
| `--stage <dir>` | Transform only one stage |
| `--force` | Overwrite modified (dirty) stage directories |

**`crane apply` — behavioral changes:**

* Renders output with `kubectl kustomize`
* Discovers stages automatically from directory structure
* Supports same stage-selection flags as `transform`
* Preflight checks: validates `kubectl` availability and `kustomization.yaml` existence before execution

> * Idempotent by design: `crane apply` delegates to `kubectl apply` with declarative semantics, so repeated runs produce the same result without side effects. Any future additions to the apply flow (pre/post hooks, logging, etc.) must preserve this property.

Reviewer: Does this mean that `crane apply` executes both `kubectl kustomize` and `kubectl apply`? Why?

Reviewer 2: No, this is a typo, thanks for reporting, should be "delegates to `kubectl kustomize`".

#### Plugin Compatibility

Existing plugins are unaffected. They continue to return JSONPatch operations via the existing plugin contract. Crane converts these operations into Kustomize Strategic Merge Patches or JSON6902 patch files during the serialization phase. This conversion is internal to `crane-lib`.

#### Stage Discovery

> Stages are discovered by convention: Crane scans the transform directory for subdirectories matching `<number>_<pluginName>`. When no explicit configuration is provided, the KubernetesPlugin is assigned priority 10, and remaining plugins are assigned priorities starting at 20 in alphabetical order, with gaps of 10 to allow user-defined custom stages.

Reviewer: Does this mean that users have to put together an entire directory structure by themselves in order to configure the transformations to be executed?

Reviewer 2: The directory structure is created with `crane transform` commands, user don't have to use commands like mkdir, etc. User can run multiple `transform` commands with for different stages (one line with comand, one stage).

```
crane transform      # uses default Kubernetes Plugin to create 10_KubenetesPlugin, typical (but not required) first step
crane transform --stage 20_SomePlugin
crane transform --stage 50_CustomEdits   
```
Then manual custom changes in stage 50_... and `crane apply` to get the final output.

Reviewer 3: So if a user has custom plugins and they want `transform` to always run it they use the existing `plugin-manager` to add these so they don't need to chain a bunch of `--stage` commands each time?

Reviewer 2: Existing plugin manager could be used, but this implementation introduced rule _one plugin per stage_ (primary for better debugging), so bunch of `transform --stage` are needed (or the migration description file is needed).

Author: After discussing it we agreed on updating the transform with multiple stages generated by a single command and add simple transform descriptor file. Tracked in https://github.com/migtools/crane/issues/370

### Upgrade / Downgrade Strategy

**Upgrade:** Users running `crane transform` after upgrading will see the new Kustomize overlay output instead of `transform-*.json` files. Result manifests will still get into the output directory.

**Downgrade:** Users downgrading to a pre-enhancement version of Crane must delete their transform directory and re-run `crane transform`, as the old version cannot read the new layout.

## Implementation History

* **2026-03:** [Issue #207](https://github.com/migtools/crane/issues/207) opened with detailed proposal for Kustomize-native output and multi-stage pipeline.
* **2026-03:** [Multi-Stage Transform Specification](https://github.com/aufi/move-crane/blob/main/drafts/spec-transform-multistep-kustomize/SUMMARY.md) drafted with full on-disk layout, CLI spec, and implementation breakdown.
* **2026-04 - present:** Implementation - open for testing, feedback and change requests.

## Drawbacks

* **Breaking change.** Removing `transform-*.json` files will break any user scripts or CI pipelines that parse the old format. This is mitigated by documentation and a deprecation period, but is still disruptive.

> * **kubectl dependency.** Adding a hard runtime requirement on `kubectl` increases the tool's prerequisites. Users in air-gapped environments must ensure `kubectl` is available.

Reviewer 3: Is there no way to consume this functionality as a library? I think it would be a mistake to require users to need `kubectl` as a hard requirement (though for most users it's probably not a huge ask)

Reviewer 2: It is technically possible import kustomize as go dependency to crane and run kustomize transformations within crane itself. Downsides are large list of added dependencies (possible conflicts primary in client-go and kyaml) and large size of the produced binary.

For me it looks cleaner and less tricky to not embed kustomize to crane, but both options are possible.

Author: In a discussion we had on this it was agreed to evaluate embedding Kustomize directly into the Crane binary to remove the kubectl dependency on the user's system. This is tracked in https://github.com/migtools/crane/issues/369

* **Complexity.** The multi-stage pipeline adds conceptual and implementation complexity. For users running a single plugin, the output still uses the staged directory layout (e.g., `10_KubernetesPlugin/`) for consistency, but only a single stage directory is created.

## Alternatives

> 1. **Embed a Kustomize library instead of shelling out to kubectl.** Pro: removes the `kubectl` runtime dependency. Con: Kustomize's Go library API is not stable and has historically been difficult to vendor. The `kubectl kustomize` path is simpler and always tracks the user's installed version.

Reviewer 3: I would push back on this; we control the version we pull in and kubectl already pulls this in as a dep, so we can just map to the latest version that kube ships.

Reviewer 2: I'm not sure about overhead with linking correct downstream kubectl/oc dependency according to given release, separated binaries don't have to solve this, but keeping this as a right question to be discussed.

Put together few notes at https://github.com/aufi/move-crane/blob/main/notes/kubectl-call-or-embed.md

Author: In a discussion we had on this it was agreed to evaluate embedding Kustomize directly into the Crane binary to remove the kubectl dependency on the user's system. This is tracked in https://github.com/migtools/crane/issues/369

    * **Decision:** Defer to a future enhancement. Kustomize components and custom transform stages cover most use cases today. Helm integration adds complexity and requires users to have Helm experience. If needed, users can use Kustomize's built-in `helmCharts` generator in their custom stages without crane changes.
4. Is it right to assume:
    1. that user-created scripts around the previous `transform-*.json` files are not supported anymore? (not affecting plugins)
        * **Decision:** Yes — `transform-*.json` files are replaced by the Kustomize overlay layout. Existing user scripts that parse `transform-*.json` will break. Document the migration path and consider a one-release deprecation period (already noted in Breaking Change section).
    2. that `crane apply` can call `kubectl kustomize` internally (or `oc`)? The binary itself will not have kustomize evaluation logic embedded.
        * **Update:** Embedding Kustomize directly into the Crane binary is being evaluated to remove the `kubectl` runtime dependency ([crane#369](https://github.com/migtools/crane/issues/369)). The current implementation shells out to `kubectl kustomize`; this may change.

>         * **Decision:** Currently yes — `crane apply` shells out to `kubectl kustomize`. Embedding Kustomize as a Go library is being evaluated ([crane#369](https://github.com/migtools/crane/issues/369)) to remove the `kubectl` runtime dependency. The answer may change based on that evaluation.

Reviewer 4: ```suggestion
        * **Decision:** Resolved — Kustomize is now embedded directly into the Crane binary as a Go library (`sigs.k8s.io/kustomize/api/krusty`). There is no runtime dependency on `kubectl` or `oc` for transform or apply operations ([crane#369](https://github.com/migtools/crane/issues/369), [crane#388](https://github.com/migtools/crane/pull/388)).
```

This creates several real problems:

* **No standard customization by users.** Teams that want to layer additional changes (e.g., injecting a sidecar, overriding resource limits) must do so outside the Crane workflow, losing the single-pipeline benefit.
* **No traceability for multiple plugins.** When multiple plugins run, their patches are merged into a single file per resource. If a transformation produces unexpected results, there is no way to determine which plugin was responsible without re-running plugins individually.

### Why Now

The earlier [kustomize-support](https://github.com/konveyor/enhancements/tree/master/enhancements/crane-2.0/kustomize-support) enhancement was rejected because it offered "no more than `kustomize init --autodetect`." Since then, the plugin system has matured, the need for traceability has intensified (users run 3-5 plugins in production pipelines), and a detailed specification for multi-stage execution has been drafted. This proposal addresses the deeper integration that the earlier enhancement deferred.

### Goals

> 1. **Kustomize-native output.** `crane transform` produces a valid Kustomize overlay (`kustomization.yaml` + patch files) instead of flat JSONPatch files. Users can run `kubectl kustomize <transform-dir>` to preview the final manifests at any time.

Reviewer 4: ```suggestion
1. **Kustomize-native output.** `crane transform` produces a valid Kustomize overlay (`kustomization.yaml` + patch files) instead of flat JSONPatch files. Crane renders these internally via the embedded kustomize library. Users can also run `kubectl kustomize <transform-dir>` externally to preview the final manifests.
```

2. **Multi-stage pipeline.** Each plugin produces its own isolated stage directory (`10_KubernetesPlugin/`, `20_OpenShiftPlugin/`, etc.). Each stage is independently inspectable and runnable. Each stage is suited for actual input manifests (from export or previous stage output), this means a new stage will internally apply the previous one.
3. **Debug-friendly stages.** Each stage (plugin and custom) produces its own input and output manifests under directory `transform/.work/10_ThePlugin`.
4. **Deterministic, Git-friendly output.** Patch files use stable naming (`<namespace>--<group>-<version>--<Kind>--<name>.patch.yaml`) and deterministic ordering so that repeated runs produce identical output when input is unchanged.
5. **Zero plugin changes.** Existing plugins continue to return JSONPatch operations. Crane handles the conversion to Kustomize patches transparently.
6. **Standard toolchain integration.** The output works with `kubectl`, `kustomize`, ArgoCD, Flux, OPA/Gatekeeper, and other tools that consume Kustomize overlays.

### Non-Goals

> * **Replacing Kustomize.** Crane is not building its own overlay engine. It produces standard Kustomize artifacts and delegates rendering to `kubectl kustomize`.

Reviewer 4: ```suggestion
* **Replacing Kustomize.** Crane is not building its own overlay engine. It produces standard Kustomize artifacts and renders them via the embedded kustomize library.
```

* **Plugin distribution.** How plugins are discovered, installed, and updated is covered by the existing crane [plugin-management](https://github.com/konveyor/enhancements/tree/master/enhancements/crane-2.0/plugin-management) system.
* **Operator/controller integration.** This enhancement covers the CLI workflow only. Integration with the Crane operator or mig-controller is out of scope.

> * **Semantic validation between stages.** Cross-stage content validation is delegated to `kubectl kustomize` at apply time.

Reviewer 4: ```suggestion
* **Semantic validation between stages.** Cross-stage content validation is delegated to the embedded kustomize at apply time.
```

## Proposal

### User Stories

#### Story 1: Transformation with the Default Plugin

**As a** platform engineer migrating an application, **I want to** run `crane transform` to execute the default `10_KubernetesPlugin`, **So that** server-managed metadata is removed, generating standard Kustomize patches that I can easily inspect and debug.

* **Patch Generation:** The plugin automatically generates JSONPatch operations that Crane transparently converts into Kustomize patches, saving them in `transform/10_KubernetesPlugin/patches/` using a predictable, Git-friendly naming convention (`<namespace>--<group>-<version>--<Kind>--<name>.patch.yaml`).

> * **Inspection:** At any time, I can preview the final cleaned YAML before applying it to the cluster by simply running `kubectl kustomize transform/10_KubernetesPlugin/`.

Reviewer 4: ```suggestion
* **Inspection:** At any time, I can preview the final cleaned YAML before applying it to the cluster. Crane renders this internally, or I can run `kubectl kustomize transform/10_KubernetesPlugin/` externally for manual inspection.
```

* **Debugging:** If I need to see exactly what went into the transformation or what came out, I can check the intermediate raw snapshots inside the `transform/.work/10_KubernetesPlugin/input/` and `output/` directories.

#### Story 2: Multi-Stage Transformation with Multiple Existing Plugins

**As a** platform engineer performing a cross-platform migration (such as moving from OpenShift to vanilla Kubernetes), **I want to** run a multi-stage pipeline where each plugin has its own isolated directory (e.g., `10_KubernetesPlugin` and `20_OpenShiftPlugin`), **So that** I can accurately trace patch generation across plugins, inspect intermediate states, and debug unexpected changes without dealing with a single massive patch file.

* **Patch Generation:** Each plugin's output is isolated into its own stage directory with its own `kustomization.yaml` and `patches/` folder, meaning patches from different plugins are never merged into an opaque file.

> * **Inspection:** I can inspect the output of any specific stage individually using standard tools like `kubectl kustomize <stage-dir>`.

Reviewer 4: ```suggestion
* **Inspection:** I can inspect the output of any specific stage individually. Crane renders each stage internally, or I can use `kubectl kustomize <stage-dir>` externally for manual verification.
```

* **Debugging:** If a resource (like a Route or Deployment) is incorrectly modified, I can easily isolate the behavior of a single plugin by inspecting its specific patch folder or by re-executing just that specific stage using the `crane transform --stage <stage-dir>` command.

#### Story 3: Custom Transformation Stage with User-Created Modifications

**As a** platform engineer adapting migrated resources to new environment standards, **I want to** append a custom transformation stage (e.g., `50_CustomModifications`) to the pipeline, **So that** I can layer GitOps-friendly declarative patches and overrides (like swapping image tags or updating namespaces) directly on top of the automated Crane output.

* **Modification Generation:** I can manually change the `kustomization.yaml` inside my custom stage directory and define custom modifications, like `namespace: migrated-app` or an `images:` block to swap container image tags, or changing annotation labels, or write custom patches without needing to write or compile a Go-based Crane plugin.

> * **Inspection:** I can verify my custom changes alongside the automated ones by running `kubectl kustomize` on my custom stage directory, ensuring my edits correctly apply over the output of the previous stage.

Reviewer 4: ```suggestion
* **Inspection:** I can verify my custom changes alongside the automated ones by running `crane apply` or `kubectl kustomize` on my custom stage directory, ensuring my edits correctly apply over the output of the previous stage.
```

  apply           Apply the transformations to the exported resources and save results in an output directory
  completion      Generate the autocompletion script for the specified shell
  convert         Convert a deprecated resource to its replacement
  export          Export the namespace resources in an output directory
  help            Help about any command
  plugin-manager  Plugin-manager is command that helps manage plugins
  skopeo-sync-gen Generate source yaml for skopeo sync and write the result to stdout
  transfer-pvc    transfer a pvc data from one kube context to another
  transform       Create the transformations for the exported resources and plugins and save the results in a transform directory
  tunnel-api      set up an openvpn tunnel to access an (source) on premise cluster from a (cloud) destination cluster
  version         Return the current crane (and crane-lib) version
```

This enhancement focuses on `transform` and `apply`.

**`crane transform` — new flags:**

| Flag | Purpose |
|------|---------|
| `--stage <dir>` | Transform only one stage |
| `--force` | Overwrite modified (dirty) stage directories |

**`crane apply` — behavioral changes:**

> * Renders output with `kubectl kustomize`

Reviewer 4: ```suggestion
* Renders output via embedded kustomize library (no external `kubectl` or `oc` dependency)```

* Discovers stages automatically from directory structure
* Supports same stage-selection flags as `transform`
* Preflight checks: validates `kubectl` availability and `kustomization.yaml` existence before execution

> * Deterministic by design: `crane apply` delegates manifest rendering to `kubectl kustomize`, so repeated runs with unchanged inputs produce identical rendered output. Any future additions to the apply flow (pre/post hooks, logging, etc.) must preserve this property.

Reviewer 4: ```suggestion
* Preflight checks: validates `kustomization.yaml` existence before execution
* Deterministic by design: `crane apply` renders manifests via the embedded kustomize library, so repeated runs with unchanged inputs produce identical rendered output. Any future additions to the apply flow (pre/post hooks, logging, etc.) must preserve this property.
```

#### One Plugin Per Stage

Each stage executes exactly one plugin. This isolation ensures that when a transformation produces unexpected results, the responsible plugin can be identified immediately by inspecting the corresponding stage directory. Users who need multiple plugins chain multiple `crane transform --stage` invocations (one per plugin). For recurring multi-plugin workflows, a migration descriptor file is planned ([crane#370](https://github.com/migtools/crane/issues/370)) to allow defining all stages in a single configuration.

#### Determinism

For stable Git diffs and reproducible CI runs:

* Resources within type files preserve creation order
* `resources` and `patches` lists in `kustomization.yaml` are sorted lexicographically by filename
* Patch filenames follow the pattern `<namespace>--<group>-<version>--<Kind>--<name>.patch.yaml`
* Stage directories follow `<priority>_<pluginName>`

### Security, Risks, and Mitigations

> **kubectl dependency.** `crane apply` currently delegates to `kubectl kustomize` at runtime. Embedding Kustomize as a Go library is being evaluated ([crane#369](https://github.com/migtools/crane/issues/369)) to remove this external dependency. Until then, a preflight check verifies `kubectl` availability with a clear error message.

Reviewer 4: ```suggestion
**Kustomize embedding.** Kustomize is embedded directly into the Crane binary as a Go library (`sigs.k8s.io/kustomize/api/krusty`). There is no runtime dependency on `kubectl` or `oc` for transform or apply operations ([crane#369](https://github.com/migtools/crane/issues/369), [crane#388](https://github.com/migtools/crane/pull/388)).
```

**Stage state drift.** Plugin-backed stages (e.g., `10_KubernetesPlugin`) can be safely regenerated by `crane transform`. User-modified custom stages require `--force` to overwrite. (See Open Question 2 for full policy discussion.)

**Breaking change for existing users.** The `transform-*.json` file layout is removed. Mitigation: document the migration path; consider a one-release deprecation period where both formats are emitted.

## Design Details

### Implementation Epics

| Epic | Scope | Key Deliverables |
|------|-------|------------------|
| **A — crane-lib: Kustomize Foundations** | Library | `TransformArtifact`/`PatchTarget` types, Kustomize serializer, resource grouping, report structs |
| **B — crane: Transform Refactor** | CLI | Kustomize overlay generation, resource type files, deterministic ordering, dirty check |
| **C — Multi-Stage Pipeline** | CLI + Lib | Stage discovery, per-stage execution, stage-aware CLI flags, priority auto-assignment |

> | **D — crane: Apply Refactor** | CLI | `kubectl kustomize` delegation, preflight checks, output handling |

Reviewer 4: ```suggestion
| **D — crane: Apply Refactor** | CLI | Embedded kustomize rendering, output handling |
```

### Upgrade / Downgrade Strategy

**Upgrade:** Users running `crane transform` after upgrading will see the new Kustomize overlay output instead of `transform-*.json` files. Result manifests will still get into the output directory.

**Downgrade:** Users downgrading to a pre-enhancement version of Crane must delete their transform directory and re-run `crane transform`, as the old version cannot read the new layout.

## Implementation History

* **2026-03:** [Issue #207](https://github.com/migtools/crane/issues/207) opened with detailed proposal for Kustomize-native output and multi-stage pipeline.
* **2026-03:** [Multi-Stage Transform Specification](https://github.com/aufi/move-crane/blob/main/drafts/spec-transform-multistep-kustomize/SUMMARY.md) drafted with full on-disk layout, CLI spec, and implementation breakdown.
* **2026-04 - present:** Implementation - open for testing, feedback and change requests.

## Drawbacks

* **Breaking change.** Removing `transform-*.json` files will break any user scripts or CI pipelines that parse the old format. This is mitigated by documentation and a deprecation period, but is still disruptive.

> * **kubectl dependency (under evaluation).** The current implementation requires `kubectl` at runtime. Embedding Kustomize as a library is being evaluated ([crane#369](https://github.com/migtools/crane/issues/369)) to eliminate this prerequisite.

Reviewer 4: ```suggestion
* **~~kubectl dependency.~~** Resolved — Kustomize is now embedded. No `kubectl` runtime dependency for transform or apply ([crane#369](https://github.com/migtools/crane/issues/369), [crane#388](https://github.com/migtools/crane/pull/388)).
```

* **Complexity.** The multi-stage pipeline adds conceptual and implementation complexity. For users running a single plugin, the output still uses the staged directory layout (e.g., `10_KubernetesPlugin/`) for consistency, but only a single stage directory is created.

## Alternatives

> 1. **Embed a Kustomize library instead of shelling out to kubectl.** Pro: removes the `kubectl` runtime dependency. Con: Kustomize's Go library API is not stable and has historically been difficult to vendor. **This option is being actively evaluated** ([crane#369](https://github.com/migtools/crane/issues/369)).

Reviewer 4: ```suggestion
1. **~~Embed a Kustomize library instead of shelling out to kubectl.~~** **Adopted.** Kustomize is now embedded via `sigs.k8s.io/kustomize/api/krusty` ([crane#369](https://github.com/migtools/crane/issues/369), [crane#388](https://github.com/migtools/crane/pull/388)). The `kustomize/api` module was already a transitive dependency, so binary size did not increase.
```

2. **Keep JSONPatch files but add a `crane render` command.** Pro: no breaking change. Con: still requires users to learn Crane-specific tooling; does not integrate with the broader Kustomize/GitOps ecosystem.
3. **Use Helm charts instead of Kustomize overlays.** Pro: Helm provides great flexibility with templating system and it is widely adopted. Con: Typical k8s workload migration process aims to clean resources with minimal needed changes to make possible to be imported to the target cluster, templating should not be required in most cases. Helm's templating model is fundamentally different from patch-based transformation.

## Infrastructure Needed

* No new repositories required. Changes land in [migtools/crane](https://github.com/migtools/crane) and [konveyor/crane-lib](https://github.com/konveyor/crane-lib).

> * CI must have `kubectl` available for integration tests (already the case in the existing test infrastructure).

Reviewer 4: ```suggestion
* CI must have `kubectl` available for E2E tests that deploy to clusters (already the case in the existing test infrastructure). Transform and apply unit tests do not require `kubectl`.
```
